frappe/model/db_query.py

```diff
@@ -275,10 +275,32 @@ def prepare_args(self):
 		# query dict
 		args.tables = self.tables[0]
 
-		# left join parent, child tables
+		user_permissions = {}
+
+		if self.tables[1:]:
+			# Add a condition to check if the child table exists; otherwise, it may result in a maximum recursion error if the cache doesn't exist. Also, adding this inside the loop doesn't seem ideal.
+			user_permissions = frappe.permissions.get_user_permissions(self.user)
+
 		for child in self.tables[1:]:
+			child_link_fields = frappe.get_meta(child.replace("`tab", "").replace("`", "")).get_link_fields()
+
+			child_conditions = ""
+
+			for child_df in child_link_fields:
+				if child_df.get("ignore_user_permissions"):
+					continue
+				user_permission_values = user_permissions.get(child_df.get("options"), {})
+				docs = [
+					p["doc"]
+					for p in user_permission_values
+					if not p.get("applicable_for") or p.get("applicable_for") == self.doctype
+				]
+				if docs:
+					values = ", ".join(frappe.db.escape(d, percent=False) for d in docs)
+					child_conditions += f" AND {child}.`{child_df.fieldname}` IN ({values})"
```

Comment:
This shouldn't be copy-pasted. Write/reuse some abstraction that's present in `add_user_permissions`